ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java

```diff
@@ -123,15 +123,20 @@ public boolean isOpened() {
    * Update the last committed index.
    * @param majorityIndex the index that has achieved majority.
    * @param currentTerm the current term.
+   * @param isLeader Is this server the leader?
    * @return true if update is applied; otherwise, return false, i.e. no update required.
    */
-  public boolean updateLastCommitted(long majorityIndex, long currentTerm) {
+  public boolean updateLastCommitted(long majorityIndex, long currentTerm, boolean isLeader) {
     try(AutoCloseableLock writeLock = writeLock()) {
       final long oldCommittedIndex = getLastCommittedIndex();
       final long newCommitIndex = Math.min(majorityIndex, getFlushIndex());
       if (oldCommittedIndex < newCommitIndex) {
-        // Only update last committed index for current term. See §5.4.2 in
-        // paper for details.
+        if (!isLeader) {
```

Comment:
Thanks for the quick fix !

There is a subtle thing here: it should be min of leaderCommit and index of the last new appended entry from this AppendEntriesRequest.

<img width="389" alt="image" src="https://user-images.githubusercontent.com/8412672/99776540-f5202c00-2b4b-11eb-9bf7-a9469961dbe9.png">

Here is the fault scenario:

Given the format of each log entry is <term, index>, 

follower has raft log as
``
<1,1>, <1,2>, <1,3>, <1,4>, <1,5>, <1,6>, 
``
Its flush index is 6, commit index is 3 (which is get from leader of term 1)

leader has raft log as
``
<1,1>, <1,2>, <1,3>, <2,4>, <2,5>
``
Say, its commit index is 4

leader send an empty heartbeat with <previousTerm, previousIndex> as <1,3>, which can pass consistency check.

In this case, follower will advance its commit index to 4, but actually `<1,4>` is not a new added log entry.

Reply:
<img width="646" alt="截屏2020-11-20 下午4 32 16" src="https://user-images.githubusercontent.com/8412672/99777973-0702ce80-2b4e-11eb-9505-466c122a09dc.png">

https://github.com/logcabin/logcabin/blob/master/Server/RaftConsensus.cc

This is how the author of raft handle this issue: he shrinks the commit index that it sends to the followers.

For above heartbeat case, since no entries is packed into the append entries request, the commit index received by the follower is just preLogIndex.

Reply:
Thanks a lot for the hint!  It would take me a long time to figure it out.

Reply:
@szetszwo Minor comment: instead of sending a truncated leaderCommit, should we use minimum of leader commit and entry size in appendEntriesAsync itself?


      private CompletableFuture<AppendEntriesReplyProto> appendEntriesAsync(
      RaftPeerId leaderId, long leaderTerm, TermIndex previous, long leaderCommit, long callId, boolean initializing,
      List<CommitInfoProto> commitInfos, LogEntryProto... entries) throws IOException {
      ...
        long commitIndex = min(leaderCommit, previous.getIndex() + entries.length);
        state.updateCommitIndex(commitIndex, currentTerm, false);
      ...

Reply:
Sure, it is a good idea to also check it at follower side.